test/e2e/main_test.go

```diff
@@ -21,66 +21,142 @@ import (
 	"testing"
 
 	operatorFramework "github.com/coreos/prometheus-operator/test/framework"
-)
 
-var framework *operatorFramework.Framework
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+)
 
-// Basic set of e2e tests for the operator:
-// - config reload (with and without external url)
+var (
+	framework *operatorFramework.Framework
+	opImage   *string
+)
 
 func TestMain(m *testing.M) {
-	kubeconfig := flag.String("kubeconfig", "", "kube config path, e.g. $HOME/.kube/config")
-	opImage := flag.String("operator-image", "", "operator image, e.g. quay.io/coreos/prometheus-operator")
-	ns := flag.String("namespace", "prometheus-operator-e2e-tests", "e2e test namespace")
+	kubeconfig := flag.String(
+		"kubeconfig",
+		"",
+		"kube config path, e.g. $HOME/.kube/config",
+	)
+	opImage = flag.String(
+		"operator-image",
+		"",
+		"operator image, e.g. quay.io/coreos/prometheus-operator",
+	)
 	flag.Parse()
 
 	var (
 		err      error
 		exitCode int
 	)
 
-	if framework, err = operatorFramework.New(*ns, *kubeconfig, *opImage); err != nil {
+	if framework, err = operatorFramework.New(*kubeconfig, *opImage); err != nil {
 		log.Printf("failed to setup framework: %v\n", err)
 		os.Exit(1)
 	}
 
-	defer func() {
-		if exitCode != 0 {
-			if err := framework.PrintEvents(); err != nil {
-				log.Printf("failed to print events: %v", err)
-			}
-			if err := framework.PrintPodLogs(framework.Namespace.Name, framework.OperatorPod.Name); err != nil {
-				log.Printf("failed to print Prometheus Operator logs: %v", err)
-			}
-		}
+	exitCode = m.Run()
 
-		if err := framework.Teardown(); err != nil {
-			log.Printf("failed to teardown framework: %v\n", err)
-			exitCode = 1
-		}
+	os.Exit(exitCode)
+}
 
-		os.Exit(exitCode)
-	}()
+// TestAllNS tests the Prometheus Operator watching all namespaces in a
+// Kubernetes cluster.
+func TestAllNS(t *testing.T) {
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup(t)
 
-	exitCode = m.Run()
+	ns := ctx.CreateNamespace(t, framework.KubeClient)
+
+	err := framework.CreatePrometheusOperator(ns, *opImage, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// t.Run blocks until f returns or calls t.Parallel to become a parallel
```

Comment:
> cks until f return
What do you mean with f? Framework?

Reply:
`f` is the function passed as the second argument to `t.Run`. Good question. I have adjusted the description.